### PR TITLE
Add Orleans.Serialization library as a high-fidelity, version-tolerant serializer

### LINKED ISSUE #2653: Version-Tolerant Serialization
Our current serializer is very fast & flexible, but it does not offer any guarantees about tolerance as users change data types. Currently, the advice is that users should use a version-tolerant serializer such as Bond or Protocol Buffers if that's required (as it is in upgrade scenarios and for storage). These serializers require that users specify a unique id for each field/property. They have the following downsides:
* Users must either not use generic/ambiguous fields, such as `object` /`SomeBaseClass`/`ISomeInterface` where a concrete class will be used at runtime, or user application-specific logic to read those fields and deserialize them into the correct type.
* Read-only/Get-only/inaccessible fields cannot be used, making it difficult to support immutable data structures.
* In the case of Protocol Buffers, the user also needs to use an IDL file (ending in .proto) to describe each message. There is no official MSBuild Protocol Buffers integration. Bond has an optional IDL.
* The C# Protocol Buffers implementation does not seem to support round-tripping a version-tolerant message with full fidelity. Bond does, but only if the field is of type `IBonded<T>` (which is basically an envelope type which defers deserialization of its payload)

Those serializers do offer cross-language support, whereas ours does not.

Our serializer is special in that is supports:
* Ambiguous types such as `object`
* Generic types
* Read-only/Get-only/inaccessible fields
* Minimal fuss, so users can serialize just about any POCO without additional markup.

The question is, can we retain some of this low-fuss nature while also improving robustness with version-tolerance? I believe we can while adding only minimal fuss.

The benefits are clear: much greater ability to reason about upgrade scenarios and data integrity.

These are the changes I would propose if we were to implement this:
* Type serialization needs to be improved so that types can have fixed identifiers, where the base type is comprised of two parts: `AssemblyId : ushort` & `TypeId : ushort`.
* Serialization for generics & array types would need to be updated
* `AssemblyProcessor` would need to scan all `[assembly: AssemblyId(id)]` & `[TypeId(id)]` attributes. It should throw if multiple assemblies have the same type id.
* A `[FieldId(id)]` attribute would be created.
* A variant/replacement for the ILBasedSerializer would be created where deserializers follow a pattern similar to this very rough pseudo-code, which is quite similar to the generated ProtoBufs code:
```C#
length = input.ReadInt();
start = input.Position;

while (consumed < length)
{
  var fieldId = input.ReadFieldId();
  switch (fieldId)
  {
    case <field1 fieldId>:
      result.field1 = input.<ReadField1Type>(); // For primitive types (we do this today)
      break;
    case <field2 fieldId>:
      result.field2 = serializationManager.DeserializeInner(input, typeof(field2Type)) // For complex types.
      break;
    default:
      // If result type has field with '[ExtensionData]'
        if (result.<extensionDataField> == null) result.<extensionDataField> = new ExtensionData();
        result.<extensionDataField>.Add(tag, input.ReadUnknown())
      // Else, if the type does not support extension data
        input.ReadUnknown();
      break;
  }

  consumed = input.Position - start;
}

byte[] ReadUnknown(input)
{
  // Find length of unknown
  // Copy & return <length> raw bytes

  var typeTag = input.PeekTypeTag()
  byte[] result;
  switch (typeTag)
  {
    case TypeTag.LengthPrefixed: // Complex type
    {
      var length = input.PeekInt(offset: sizeof(TypeTag));
      return  input.ReadRaw(sizeof(TypeTag) + sizeof(int) + length);
    }
    case TypeTag.Int32:
     return input.ReadRaw(sizeof(TypeTag) + sizeof(int))
// ... other primitive types
  }
}
```
The code for serializers is conceptually similar, but simpler.

A version-tolerant object would then look like this:
```C#
[TypeId(41592714)] // Omitting this means that the type would be serialized using its full name.
public class MyType<T>
{
  [FieldId(1)]
  public int MyInt { get; set; }
  
  [FieldId(2)]
  public string MyString { get; set; }

  [FieldId(3)]
  public T SomeOtherType { get; set; }
}
```

* We could serialize field ids using variable-length integers to save space, but that's an optimization.

In the past I've suggested that we shouldn't invest in these changes, but my mind has been slowly changing over time for a few reasons:
* Experience with Bond & Protocol Buffers shows that they add significant friction to the development progress. Fundamentally, those serializers don't offer the ease-of-use I'd like. One of our goals is to reduce this friction wherever possible.
* We've been making improvements in our own serializers, such as making SerializationManager non-static and the IL-based serializer, which has had time to stabilize
* I see a need for version-tolerant serializers internally, with pressure to alter some of our core types to support new features (eg, opening GrainId, making it generic, improvements to GrainReference & Message).

What are your thoughts?

### LINKED ISSUE #5021: The new code generator adds everything to feature.AddKnownType
This was a problem with the old code generator already as documented in #4437, but with the new code generator it's a lot worse.

For example I have a functional tests project that has a single grain interface with a few simple calls, it references various other projects. Both code generators write about 200 lines of code for the grain method invoker, but then they generate endless `feature.AddKnownType` lines - the old one about 700 lines, the new one about **26000** lines! The generated assembly grows from 300KB to 6MB and is really slow to start up.

What's the purpose of `feature.AddKnownType`? It looks like it adds every type that is visible to Roslyn during codegen or something similar. And the added information doesn't look useful at all - it's just a list of type names.

### PR DESCRIPTION
**Pull request by ReubenBond**
This PR integrates a new version-tolerant serializer and RPC primitives into Orleans. For more context, and prior discussion, please see #2653 and [https://github.com/ReubenBond/Hagar](https://github.com/ReubenBond/Hagar).

The serializer implemented in this PR requires that developers annotate their types with an attribute on the type and an attribute on each serialized field/property. **There is an included code analyzer to ease the process, generating the per-data-member attributes.** Here is an example of an annotated type:

``` C#
[GenerateSerializer]
public class ClassWithEnumTestData
{
    [Id(0)]
    public TestEnum EnumValue { get; set; }

    [Id(1)]
    public CampaignEnemyTestType Enemy { get; set; }
}

```

The wire format takes inspiration from Bond, and Protocol Buffers, but adds support for nominal types and references, which would otherwise need to be encoded on top of the base protocol. This is the primary reason for its existence, since it is believed that this greatly improves the experience for .NET developers, giving them substantially more freedom when designing their applications and modelling their data.

Developers can configure the system to use *Newtonsoft.Json* or another package as the serializer for their types instead, similarly to how `IExternalSerializer` works today. Unlike the existing serializer, even data serialized by external serializers must conform to a basic wire format (typically, by emitting a length-prefixed sequence).

Deep copiers are decoupled from serializers, which is an issue with the existing system: developers often do not want to replace copiers when replacing serializers.

Previously unserializable/undeserializable Exceptions can be serialized with a decent degree of fidelity using this serializer, so that the receiver can see the exception type, message, stack trace, and other core properties even if the concrete exception type is not available. If the concrete exception type is not available, then the properties are deserialized into a special-purpose `UnavailableExceptionFallbackException` object.

A new NuGet package, named *Orleans.Sdk*, takes the place of installing multiple NuGet packages on grain interface/flass/

Application Parts are removed entirely: the code generator package generates attributes which serve the purpose of Application Parts in identifying the closure of assemblies which contain types that Orleans must know about. There is still a way to remove grain classes & interfaces programmatically, by configuring `GrainTypeOptions`, which has a `Classes` and an `Interfaces` property. Assemblies can also be manually included in the closure by adding an assembly-level attribute to the main assembly.

While Orleans serialization is not intended for use between untrusted parties, this serializer greatly improves security aspects of serialization in a variety of ways:

- Only known types are allowed to be serialized - types are disabled by default. Types which are annotated are allowed to be deserialized, and predicates for which types are allowed can be specified at configuration time.
- Length fields are validated to prevent an attacker to force the deserializing party to pre-allocate a buffer which is larger than the payload.

The serializer is also much faster than the existing serializer, making use of hardware intrinsics where available, and avoiding unnecessary allocations.

# RPC

Serialization and RPC are closely related. A large component of RPC is the serialization of method calls for later deserialization and invocation. This PR also includes new RPC primitives, which replace both `InvokeMethodRequest` and generated `IGrainMethodInvoker` implementations with generated `IInvokable` classes (two become one). There is one generated `IInvokable` class for each method on each grain interface. Generated `GrainReference` classes change minorly to make use of the new generated `IInvokable` classes.

The RPC supports efficient serialization of method arguments, avoiding most or all unnecessary allocations and avoiding unnecessary type information

As a part of a concerted effort to reduce unnecessary overheads, the RPC is customizable enough to allow us to implement *Orleans.Transactions* entirely outside of the core libraries, and this PR also includes that change. This means that non-transactional calls avoid checking for transaction contexts or requirements. It hopefully also means that we can support more experimentation in the future without requiring forking the codebase.

---

TODOs for this PR:

- [x] Implement renaming tolerance for RPC invokable types

Areas for future investigation/improvements:

- [ ]  Move existing `SerializationManager` and related types to a new package (*Microsoft.Orleans.Serialization.Legacy*)
- [ ]  Analyzer: allow customization of the serializer attributes through build time properties, just as the code generator does
- [ ] Add in-place copying optimization for RPC types, supporting `[Immutable]` on parameters and return values (via `[return: Immutable]`)
- [ ]  Explore emitting a separate, diffable database of types included in an application, so that annotations can be removed from code. This may end up being too 'magic' to be worthwhile, but we won't know for sure unless we explore it.

Fixes #2653
Fixes #5021

### CODE REVIEW

**Comment by haefele** [comment]
Hey, I know I'm not a contributor to this project, I just saw your tweet and thought: Gotta check this thing out!

So I noticed that you still have some Hagar leftovers in your filenames.

**Review by galvesribeiro — APPROVED** [review]
LGTM! Awesome! Looking forward to use it :shipit:

File: src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureBlobStorage.cs [hunk 1]
```diff
@@ -31,24 +32,24 @@ public class AzureBlobGrainStorage : IGrainStorage, ILifecycleParticipant<ISiloL
         private ILogger logger;
         private readonly string name;
         private AzureBlobStorageOptions options;
-        private SerializationManager serializationManager;
+        private Serializer serializationManager;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureBlobStorage.cs:35]: True, I'll run a find & replace. I also want to migrate to the more common .NET naming scheme and prefix all field names with underscores

[the remark this answers is not in the document]
Comment (galvesribeiro) [reply src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureBlobStorage.cs]: Yeah, That was a major drama recently...

**Pull request merged**